program/src/initialize_vault_operator_delegation_snapshot.rs

```diff
@@ -0,0 +1,156 @@
+    };
+
+    if vault_program.key.ne(&jito_vault_program::id()) {
+        msg!("Incorrect vault program ID");
+        return Err(ProgramError::InvalidAccountData);
+    }
+
+    if restaking_program.key.ne(&jito_restaking_program::id()) {
+        msg!("Incorrect restaking program ID");
+        return Err(ProgramError::InvalidAccountData);
+    }
+
+    NcnConfig::load(program_id, ncn.key, ncn_config, false)?;
+    Config::load(restaking_program.key, restaking_config, false)?;
+    Ncn::load(restaking_program.key, ncn, false)?;
+    Operator::load(restaking_program.key, operator, false)?;
+    Vault::load(vault_program.key, vault, false)?;
+
+    VaultOperatorDelegation::load(
+        vault_program.key,
+        vault_operator_delegation,
+        vault,
+        operator,
+        false,
+    )?;
+    VaultNcnTicket::load(vault_program.key, vault_ncn_ticket, vault, ncn, false)?;
+    NcnVaultTicket::load(restaking_program.key, ncn_vault_ticket, ncn, vault, false)?;
+
+    //TODO check that st mint is supported?
+    //TODO may not exist
+    //TODO what happens if a new Vault is added inbetween weight table creation and this? - There could be a count mismatch
+    if !vault_operator_delegation.data_is_empty() {
+        VaultOperatorDelegation::load(
+            vault_program.key,
+            vault_operator_delegation,
+            vault,
+            operator,
+            false,
+        )?;
+    }
+
+    let current_slot = Clock::get()?.slot;
+    let (ncn_epoch, ncn_epoch_length) =
+        load_ncn_epoch(restaking_config, current_slot, first_slot_of_ncn_epoch)?;
+
+    WeightTable::load(program_id, weight_table, ncn, ncn_epoch, false)?;
+    EpochSnapshot::load(program_id, ncn.key, ncn_epoch, epoch_snapshot, true)?;
+    OperatorSnapshot::load(
+        program_id,
+        operator.key,
+        ncn.key,
+        ncn_epoch,
+        operator_snapshot,
+        true,
+    )?;
+
+    let (vault_index, st_mint) = {
+        let vault_data = vault.data.borrow();
+        let vault_account = Vault::try_from_slice_unchecked(&vault_data)?;
+        (vault_account.vault_index(), vault_account.supported_mint)
+    };
+
+    let is_active: bool = {
+        let vault_ncn_ticket_data = vault_ncn_ticket.data.borrow();
+        let vault_ncn_ticket_account =
+            VaultNcnTicket::try_from_slice_unchecked(&vault_ncn_ticket_data)?;
+
+        let ncn_vault_ticket_data = ncn_vault_ticket.data.borrow();
+        let ncn_vault_ticket_account =
+            NcnVaultTicket::try_from_slice_unchecked(&ncn_vault_ticket_data)?;
+
+        let vault_ncn_okay = vault_ncn_ticket_account
+            .state
+            .is_active(current_slot, ncn_epoch_length);
+
+        let ncn_vault_okay = ncn_vault_ticket_account
+            .state
+            .is_active(current_slot, ncn_epoch_length);
+
+        let delegation_dne = vault_operator_delegation.data_is_empty();
+
+        vault_ncn_okay && ncn_vault_okay && delegation_dne
+    };
```

Comment:
One general thought is that we're making these instructions pretty long, and could do a better job of putting more stuff (especially common things like is_active checks) in functions that we can unit test.
Reduces burden on integration tests which are more cumbersome -- main logic should be testable in unit tests, ensure that things are wired properly with integration tests.
Though I'm happy we have a really solid integration test foundation, and this refactor would only be worth it once we try to get higher test coverage

Reply:
made todo comments